File: packages/examples/todo/backends/rust/src/commands/get.rs
use super::{not_found, ok, schema, Command};
use crate::store::TodoStore;
use afd::{failure, CommandResult};
use serde::Deserialize;
use serde_json::{json, Value};
use std::sync::Arc;

pub struct Get(pub Arc<TodoStore>);

#[derive(Deserialize)]
pub struct Input {
    id: String,
}

impl Command for Get {
    type Input = Input;
    const NAME: &'static str = "todo-get";
    const DESCRIPTION: &'static str = "Get a single todo by ID";
    const MUTATION: bool = false;

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": { "id": schema::id() },
            "required": ["id"]
        })
    }

    fn run(&self, input: Input) -> CommandResult<Value> {
        match self.0.get(&input.id) {
            Some(todo) => {
                let reasoning = format!("Retrieved todo \"{}\"", todo.title);
                ok(&todo, reasoning, 1.0, Vec::new())
            }
            None => failure(not_found(&input.id)),
        }
    }
}

File: packages/examples/todo/backends/typescript/src/commands/get.ts
/**
 * @fileoverview todo.get command
 */

import { defineCommand, failure, success } from '@lushly-dev/afd-server';
import { z } from 'zod';
import { store } from '../store/index.js';
import type { Todo } from '../types.js';

const inputSchema = z.object({
	id: z.string().min(1, 'Todo ID is required'),
});

export const getTodo = defineCommand<typeof inputSchema, Todo>({
	name: 'todo-get',
	description: 'Get a single todo by ID',
	category: 'todo',
	tags: ['todo', 'get', 'read', 'single'],
	mutation: false,
	version: '1.0.0',
	input: inputSchema,
	errors: ['NOT_FOUND'],

	async handler(input) {
		const todo = store.get(input.id);

		if (!todo) {
			return failure({
				code: 'NOT_FOUND',
				message: `Todo with ID "${input.id}" not found`,
				suggestion: 'Use todo.list to see available todos',
			});
		}

		return success(todo, {
			reasoning: `Retrieved todo "${todo.title}"`,
			confidence: 1.0,
		});
	},
});
